compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1465,8 +1465,10 @@ class Typer extends Namer
       case _ =>
         if tree.isInline then checkInInlineContext("inline match", tree.srcPos)
         val sel1 = typedExpr(tree.selector)
-        val selType = fullyDefinedType(sel1.tpe, "pattern selector", tree.span).widen
-
+        val rawSelectorTpe = fullyDefinedType(sel1.tpe, "pattern selector", tree.span)
+        val selType = rawSelectorTpe match
+          case c: ConstantType if tree.isInline => c
```

Comment:
It should probably be matching `Singleton` as we also do not want to widen a `x.type`.

Reply:
I updated this and added a test case